src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,963 @@
+ rowsPerImage (including { ==, > } copyExtent.height) values and minimum required bytes in copy works for every format. Also covers special code paths:
+  - bufferSize - offset < bytesPerImage * copyExtent.depth
+  - when bytesPerRow is not a multiple of 512 and copyExtent.depth > 1: copyExtent.depth % 2 {==, >} 0
+  - bytesPerRow == bytesInACompleteCopyImage
+
+* copy_with_various_offsets_and_data_sizes: test that copying data with various offset (including { ==, > } 0 and is/isn't power of 2) values and additional\
+ data paddings works for every format with 2d and 2d-array textures. Also covers special code paths:
+  - offset + bytesInCopyExtentPerRow { ==, > } bytesPerRow
+  - offset > bytesInACompleteCopyImage
+
+* copy_with_various_origins_and_copy_extents: test that copying slices of a texture works with various origin (including { origin.x, origin.y, origin.z }\
+ { ==, > } 0 and is/isn't power of 2) and copyExtent (including { copyExtent.x, copyExtent.y, copyExtent.z } { ==, > } 0 and is/isn't power of 2) values\
+ (also including {origin._ + copyExtent._ { ==, < } the subresource size of textureCopyView) works for all formats. origin and copyExtent values are passed\
+ as [number, number, number] instead of GPUExtent3DDict.
+
+* copy_various_mip_levels: test that copying various mip levels works for all formats. Also covers special code paths:
+  - the physical size of the subresouce is not equal to the logical size
+  - bufferSize - offset < bytesPerImage * copyExtent.depth and copyExtent needs to be clamped
+
+* copy_with_no_image_or_slice_padding_and_undefined_values: test that when copying a single row we can set any bytesPerRow value and when copying a single\
+ slice we can set rowsPerImage to 0. Also test setting offset, rowsPerImage, mipLevel, origin, origin.{x,y,z} to undefined.
+
+* TODO: 
+  - add another initMethod which renders the texture
+  - check that CopyT2B doesn't overwrite any other bytes
+  - because of expectContests 4-bytes alignment we don't test CopyT2B with buffer size not divisible by 4
+  - add tests for 1d / 3d textures
+`;
+
+import { params, poptions } from '../../../common/framework/params_builder.js';
+import { makeTestGroup } from '../../../common/framework/test_group.js';
+import { assert, unreachable } from '../../../common/framework/util/util.js';
+import {
+  kSizedTextureFormatInfo,
+  SizedTextureFormat,
+  kSizedTextureFormats,
+} from '../../capability_info.js';
+import { GPUTest } from '../../gpu_test.js';
+import { align } from '../../util/math.js';
+import { getTextureCopyLayout, TextureCopyLayout } from '../../util/texture/layout.js';
+
+interface TextureCopyViewWithRequiredOrigin {
+  texture: GPUTexture;
+  mipLevel: number | undefined;
+  origin: Required<GPUOrigin3DDict>;
+}
+
+type InitMethod = 'WriteTexture' | 'CopyB2T';
+type CheckMethod = 'PartialCopyT2B' | 'FullCopyT2B';
+
+// This describes in what form the arguments will be passed to WriteTexture/CopyB2T/CopyT2B.
```

Comment:
fyi: If you use jsdoc comments:
```ts
/** comment */
or
/**
 * comment
 */
```
then this will be picked up by IDEs and shown on mouseover.

Not super necessary within a file, but useful for helpers in shared files. We don't use this much in this project yet, but a few things have been updated to use it.